### PR TITLE
Use sdk level in RuntimeEnvironment to choose rumtimeadapter instand of sdk_int, fail proof for null viewRootImpl, Making a11y shadows parcelable.

### PR DESCRIPTION
**Pull request by yuzhong-google**
We have seen users trying to test on android-15 by setting Build.SDK_INT to 15 which is not supported, this is not the right way to define API level so the test can still be compiled without warning/error. This causes api level difference between the android.jar and RuntimeAdapter, which results in runtime errors. 

ActivityController setUp ViewRootImpl of a window in runPause, if a test pause thread before initializing an activity (because of #1582 for instance). ViewRootImpl will be null then we can't call dispatchResized.
@jongerrish Please take a look at the change. 

Making a11y shadows parcelable.

### CODE REVIEW

File: robolectric/pom.xml [hunk 1]
```diff
@@ -131,6 +131,9 @@
       <plugin>
         <groupId>org.apache.maven.plugins</groupId>
         <artifactId>maven-surefire-plugin</artifactId>
+        <configuration>
```
Comment (jongerrish) [inline robolectric/pom.xml]: Why is this necessary?

Comment (yuzhong-google) [inline robolectric/pom.xml]: It's not necessary, without the enlarged test memory space, robolectric sometimes throw OutofMemoryError when running install-robolectric, then I have to re-run the script and wait another 5 minutes.

File: robolectric-resources/src/main/java/org/robolectric/RuntimeEnvironment.java [hunk 2]
```diff
@@ -41,4 +43,8 @@ public static String getQualifiers() {
     public static void setQualifiers(String newQualifiers) {
```
Comment (jongerrish) [inline robolectric-resources/src/main/java/org/robolectric/RuntimeEnvironment.java]: Wouldn't it be better to have

setConfig(Config config)

then the relevant calls would be

RuntimeEnvironmnet.getConfig().getSdk()

and

RuntimeEnvironment.getConfig().getQualifiers()

Comment (yuzhong-google) [inline robolectric-resources/src/main/java/org/robolectric/RuntimeEnvironment.java]: In occasions api level is not specified, robolectric use api level in sdkConfig
qualifiers += "v" + sdkConfig.getApiLevel();

I think that's why RuntimeEnvironment keep a reference to the qualifier string instead of config.

File: robolectric/pom.xml (hunk 1 above)
Comment (erd) [inline robolectric/pom.xml]: You should also be able to set this as `MAVEN_OPTS` in your environment.

**Comment by erd** [comment]
When submitting pull requests, it's useful to state why you are proposing a given change.

Keep in mind that the user does not need to specify the sdk level in @Config. If it isn't specified, we fallback to the `targetSdkVersion` specified in the manifest, otherwise we fallback to a known sdk version. This change will break this behavior.

**Comment by yuzhong-google** [comment]
I'll modify the PR and add the problem we are trying to solve.

I don't think it change the behavior though, the way this qualifier is
defined is not changed so it still follows the api level picking flow. We
are just using the runtime api level instead of Build.SDK_INT which is less
reliable to decide which adapter to use.

On Tue, Aug 11, 2015 at 1:03 PM, Erich Douglass <notifications@github.com>
wrote:

> When submitting pull requests, it's useful to state why you are proposing
> a given change.
>
> Keep in mind that the user does not need to specify the sdk level in
> @Config <https://github.com/Config>. If it isn't specified, we fallback
> to the targetSdkVersion specified in the manifest, otherwise we fallback
> to a known sdk version. This change will break this behavior.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/robolectric/robolectric/pull/1985#issuecomment-130047084>
> .
>



-- 
Yu Zhong
Accessibility Research Group

**Comment by erd** [comment]
> Build.SDK_INT which is less reliable to decide which adapter to use

Can you explain why it is less reliable? We explicitly set this value in `RobolectricTestRunner`.

**Comment by yuzhong-google** [comment]
We saw cases where users use reflection to change the value in tests,
admittedly it's their fault and theoretically they can modify the qualifier
in RuntimeEnvironment if they want to... But I think it's less likely to
happen as SDK_INT is better known.

On Tue, Aug 11, 2015 at 1:14 PM, Erich Douglass <notifications@github.com>
wrote:

> Build.SDK_INT which is less reliable to decide which adapter to use
>
> Can you explain why it is less reliable? We explicitly set this value in
> RobolectricTestRunner.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/robolectric/robolectric/pull/1985#issuecomment-130050417>
> .
>



-- 
Yu Zhong
Accessibility Research Group

**Comment by jongerrish** [comment]
@erd the version of the RuntimeAdapter should match the aosp code that it is calling. The Build.VERSION.SDK_INT only relates to the actual code being called IF the test writer doesn't modify it using ReflectionHelpers.... and many tests do because they want to try and test 16 > api < 21. The "right" thing to do is to have RuntimeAdapter match the AOSP code that it is adapting.

**Comment by erd** [comment]
That's fair. I'm usually happy to let people shoot themselves in the foot :)

It sounds like users will be required to specify the sdk level in @Config now (or set a default in `robolectric.properties`), which is a change from the existing behavior. It would be nice if we could retain the existing behavior of defaulting to `targetSdkVersion` if no value is specified in the @Config.

**Comment by yuzhong-google** [comment]
Right, pickSdkVersion still follows the same fall back logic and throws an exception if @Config set an unsupported version. We are only changing how version of runtime adapter is chosen which happens after pickSdkVersion, there is nothing we can do about a forced Build.SDK_INT but this way we can at least ignore it.

**Comment by erd** [comment]
Thanks for the clarification!

File: robolectric-shadows/shadows-core/src/main/java/org/robolectric/shadows/ShadowAccessibilityEvent.java [hunk 3]
```diff
@@ -21,6 +24,9 @@
   private static final Map<StrictEqualityEventWrapper, StackTraceElement[]> obtainedInstances =
       new HashMap<>();
 
+  private static final SparseArray<StrictEqualityEventWrapper> orderedInstances = new SparseArray<>();
+
+  private static int sAllocationCount = 0;
```
Comment (pweaver-google) [inline robolectric-shadows/shadows-core/src/main/java/org/robolectric/shadows/ShadowAccessibilityEvent.java]: What's sAllocationCount for?

File: robolectric-shadows/shadows-core/src/main/java/org/robolectric/shadows/ShadowAccessibilityEvent.java [hunk 4]
```diff
@@ -21,6 +24,9 @@
   private static final Map<StrictEqualityEventWrapper, StackTraceElement[]> obtainedInstances =
       new HashMap<>();
 
+  private static final SparseArray<StrictEqualityEventWrapper> orderedInstances = new SparseArray<>();
```
Comment (pweaver-google) [inline robolectric-shadows/shadows-core/src/main/java/org/robolectric/shadows/ShadowAccessibilityEvent.java]: Why do we need a second copy of obtainedInstances? It seems to clean up creation from a parcel, but elsewhere you iterate over the array.

File: robolectric-shadows/shadows-core/src/main/java/org/robolectric/shadows/ShadowAccessibilityEvent.java [hunk 5]
```diff
@@ -228,4 +267,22 @@ public int hashCode() {
...
+    StrictEqualityEventWrapper wrapper = new StrictEqualityEventWrapper(realAccessibilityEvent);
+    int keyOfWrapper = -1;
+    for (int i = 0; i < orderedInstances.size(); i++) {
+      if (orderedInstances.valueAt(i).equals(wrapper)) {
+        keyOfWrapper = orderedInstances.keyAt(i);
+        break;
+      }
+    }
+    dest.writeInt(keyOfWrapper);
+  }
```
Comment (pweaver-google) [inline robolectric-shadows/shadows-core/src/main/java/org/robolectric/shadows/ShadowAccessibilityEvent.java]: If we're going to do this 100%, we should really extend ShadowAccessibilityRecord along with all of its getters and setters.

File: robolectric-shadows/shadows-core/src/main/resources/org/robolectric/shadows/ShadowAccessibilityWindowInfo.java.vm [hunk 6]
```diff
@@ -96,6 +144,24 @@ public class ShadowAccessibilityWindowInfo {
...
+
+  public void addChild(AccessibilityWindowInfo child) {
+    if (children == null) {
+      children = new LinkedList<>();
+    }
+
+    children.add(child);
+    ((ShadowAccessibilityWindowInfo) ShadowExtractor.extract(child)).parent =
+        mRealAccessibilityWindowInfo;
+  }
```
Comment (pweaver-google) [inline robolectric-shadows/shadows-core/src/main/resources/org/robolectric/shadows/ShadowAccessibilityWindowInfo.java.vm]: I believe we're missing three methods: equals, hashcode, and obtain(AccessibilityWindowInfo).

Comment (yuzhong-google) [inline robolectric-shadows/shadows-core/src/main/resources/org/robolectric/shadows/ShadowAccessibilityWindowInfo.java.vm]: Working on it.

File: robolectric-shadows/shadows-core/src/main/java/org/robolectric/shadows/ShadowAccessibilityEvent.java (hunk 3 above)
Comment (caseyburkhardt) [inline robolectric-shadows/shadows-core/src/main/java/org/robolectric/shadows/ShadowAccessibilityEvent.java]: Will now be monotonically increasing number for resolving parceled instances.

File: robolectric-shadows/shadows-core/src/main/java/org/robolectric/shadows/ShadowAccessibilityEvent.java (hunk 4 above)
Comment (caseyburkhardt) [inline robolectric-shadows/shadows-core/src/main/java/org/robolectric/shadows/ShadowAccessibilityEvent.java]: This is the SparseArray nonsense where it uses == instead of equals() for key lookup.  WAI after the removal of decrements to sAllocationCount.
